src/chocolatey/infrastructure.app/registration/SimpleInjectorContainerRegistrator.cs

```diff
@@ -0,0 +1,488 @@
+            throw new ApplicationException("The command '{0}' have been globally registered, but can not be found!".format_with(commandName));
+        }
+
+        private void register_instance(Type serviceType, Func<IContainerResolver, object> instanceAction)
+        {
+            ensure_not_built();
+
+            validate_service_registration(serviceType, serviceType, validate_multi_services: true);
+            remove_existing_registration(serviceType);
+
+            _instanceActionRegistrations.AddOrUpdate(serviceType, instanceAction, (key, value) => instanceAction);
+        }
+
+        private void register_service(Type interfaceType, Type serviceType, bool transient = false)
+        {
+            ensure_not_built();
+
+            if (!can_register_service(serviceType))
+            {
+                return;
+            }
+
+            var multiServiceAttribute = interfaceType.GetCustomAttribute<MultiServiceAttribute>();
+
+            if (multiServiceAttribute != null && multiServiceAttribute.IsMultiService)
+            {
+                add_to_multi_services(interfaceType, serviceType);
+            }
+            else
+            {
+                validate_service_registration(interfaceType, serviceType, validate_multi_services: true);
+                remove_existing_registration(interfaceType);
+
+                if (transient)
+                {
+                    _transientServices.AddOrUpdate(interfaceType, serviceType, (key, value) => serviceType);
+                }
+                else
+                {
+                    _singletonServices.AddOrUpdate(interfaceType, serviceType, (key, value) => serviceType);
+                }
+            }
+        }
+
+        private IEnumerable<string> remove_commands(Type commandType, string initialCommand)
+        {
+            var allCommandsForAttribute = commandType.GetCustomAttributes(typeof(CommandForAttribute), false).Cast<CommandForAttribute>();
+
+            foreach (var commandFor in allCommandsForAttribute)
+            {
+                var commandName = commandFor.CommandName.to_lower();
+                if (_allCommands.Contains(commandName))
+                {
+                    _allCommands.Remove(commandName);
+                }
+
+                Type tempType;
+
+                if (!commandName.is_equal_to(initialCommand) && _registeredCommands.TryRemove(commandName, out tempType))
+                {
+                    yield return commandName;
+                }
+            }
+        }
+
+        private void remove_existing_registration(Type interfaceType)
+        {
+            Type tempType;
+            Func<IContainerResolver, object> tempAction;
+            _transientServices.TryRemove(interfaceType, out tempType);
+            _singletonServices.TryRemove(interfaceType, out tempType);
+            _instanceActionRegistrations.TryRemove(interfaceType, out tempAction);
+        }

```

Comment:
The only type of logging we would probably have here is debug logging, which unfortunately won't even show up if it is added (already tried it).

It will also fail if the existing registration can't be removed (if I remember correctly). The only benefit of calling `TryRemove` is that it won't fail if the registration doesn't exist already.